google-beta/resource_compute_disk.go

```diff
@@ -46,6 +48,63 @@ func isDiskShrinkage(old, new, _ interface{}) bool {
 	return new.(int) < old.(int)
 }
 
+func customDiffComputeDiskDiskEncryptionKeys(diff *schema.ResourceDiff, meta interface{}) error {
+	oldConvenience, newConvenience := diff.GetChange("disk_encryption_key_raw")
+	oldNewField, newNewField := diff.GetChange("disk_encryption_key.0.raw_key")
+
+	// Either field has a value and then has another value
+	// We need to handle _EVERY_ ForceNew case in this diff
+	if oldConvenience != "" && newConvenience != "" && oldConvenience != newConvenience {
+		return diff.ForceNew("disk_encryption_key_raw")
```

Comment:
@rileykarson This convenience field should have been removed already in https://github.com/terraform-providers/terraform-provider-google/commit/5a2b9420bc4fdeccd76b536636c754d30654fd81 so I don't think we should have to be dealing it in the custom diffs any more.

Reply:
Ah nm, it's _going_ to be removed but this change is releasing 'before' the removal. It does seem like a lot of change to a field that's going to be removed imminently though.

Reply:
Correct! But the field didn't work properly before and we can't exactly do a deprecation from a working field -> a field that doesn't work because users don't have a chance to migrate. So we aren't able to do the removal in `2.0.0` without this change now- does that make sense?

Reply:
🤔 not particularly. The field didn't work correctly, but as of 2.0.0 the old field would be gone and there would be a new field that did work. Since it's moving from a field that wasn't working to a field that was I'm not sure how that impacts the migration story?

Reply:
So in `master`/`1.19.1` the old field works and migrating from the old field ->  new field works but creating a new resource using the new field doesn't work. We only tested creation with the old field and a migration to the new field, we didn't test creation with the new field.

To deprecate a field, we need the old and new fields to work simultaneously for at least 1 release before a major release; this change fixes the `Create` path so that we've properly deprecated the old field can actually do the removal we already did that made the new field work properly in `2.0.0`.

Like I said in another thread here - time traveling code is fun. We can talk through this offline if there's anything else you're wondering about.

Reply:
IRL conversation summarized - Riley was saying there needs to always be a point where if there is a deprecation warning that a user can change their current config and see the deprecation warning go away and experience the future functionality. This way somebody could address all the deprecation warnings display by 1.20 and be able to upgrade to 2.0 seamlessly.

Reply:
And to be 100% sure, the right move for me here is to leave this code in, right?  I thought so, and I still think so.

Reply:
Correct, leave it.